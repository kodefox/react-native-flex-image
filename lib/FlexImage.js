import * as tslib_1 from "tslib";
import React, { Component } from 'react';
// @ts-ignore
import resolveAssetSource from 'react-native/Libraries/Image/resolveAssetSource';
import { View, Text, Image, ActivityIndicator, TouchableOpacity, Animated, } from 'react-native';
export default class FlexImage extends Component {
    constructor() {
        super(...arguments);
        this._pendingGetSize = null;
        this.state = {
            isLoading: true,
            ratio: -1,
            error: '',
            thumbnailOpacity: new Animated.Value(0),
        };
        this._onThumbnailLoad = () => {
            Animated.timing(this.state.thumbnailOpacity, {
                toValue: 1,
                duration: 250,
            }).start();
        };
        this._onLoad = () => {
            Animated.timing(this.state.thumbnailOpacity, {
                toValue: 0,
                duration: 250,
            }).start();
        };
        this._onLoadSuccess = (width, height) => {
            let ratio = width / height;
            this.setState({
                isLoading: false,
                ratio,
            });
        };
        this._onLoadFail = (error) => {
            this.setState({
                isLoading: false,
                error: 'Error: ' + error.message,
            });
        };
    }
    componentDidMount() {
        let { source, thumbnail, loadingMethod } = this.props;
        let ratio = -1;
        let error = '';
        let isLoading = true;
        let src = thumbnail && loadingMethod === 'progressive' ? thumbnail : source;
        if (typeof src === 'number') {
            let imageSource = resolveAssetSource(src);
            if (imageSource) {
                let { width, height } = imageSource;
                if (width && height) {
                    ratio = width / height;
                }
            }
            else {
                error = 'Error: Failed to retrieve width and height of the image';
            }
            isLoading = false;
        }
        else {
            this._pendingGetSize = getImageSize(src, this._onLoadSuccess, this._onLoadFail);
        }
        this.setState({
            ratio,
            isLoading,
            error,
        });
    }
    componentWillUnmount() {
        if (this._pendingGetSize) {
            this._pendingGetSize.cancel();
        }
    }
    render() {
        let _a = this.props, { source, style, onPress, loadingComponent, thumbnail, loadingMethod, errorComponent } = _a, otherProps = tslib_1.__rest(_a, ["source", "style", "onPress", "loadingComponent", "thumbnail", "loadingMethod", "errorComponent"]);
        let { isLoading, ratio, error, thumbnailOpacity } = this.state;
        if (isLoading && loadingMethod !== 'progressive') {
            let loadingIndicator = loadingComponent || (<ActivityIndicator size="large"/>);
            return (<View style={[{ justifyContent: 'center', alignItems: 'center' }, style]}>
          {loadingIndicator}
        </View>);
        }
        if (error) {
            return (<View style={[{ justifyContent: 'center', alignItems: 'center' }, style]}>
          {errorComponent ? errorComponent : <Text>{error}</Text>}
        </View>);
        }
        let imageSource = null;
        if (typeof source === 'number') {
            imageSource = source;
        }
        else {
            // eslint-disable-next-line no-unused-vars
            let { uri, width, height } = source, other = tslib_1.__rest(source, ["uri", "width", "height"]);
            imageSource = Object.assign({ uri }, other);
        }
        return (<TouchableOpacity onPress={onPress} disabled={!onPress} style={[{ aspectRatio: ratio }, style]}>
        {thumbnail && loadingMethod === 'progressive' && (<Animated.Image {...otherProps} source={thumbnail} style={{
            width: '100%',
            height: '100%',
            opacity: thumbnailOpacity,
            zIndex: 1,
        }} onLoad={this._onThumbnailLoad} testID="progressiveThumbnail"/>)}
        <Animated.Image {...otherProps} source={imageSource} style={{ width: '100%', height: '100%', position: 'absolute' }} onLoad={this._onLoad}/>
      </TouchableOpacity>);
    }
}
// A cancellable version of Image.getSize
export function getImageSize(source, onSuccess, onFail) {
    let isCancelled = false;
    Image.getSize(source.uri, (width, height) => {
        if (!isCancelled) {
            onSuccess(width, height);
        }
    }, (error) => {
        if (!isCancelled) {
            onFail(error);
        }
    });
    return {
        cancel: () => {
            isCancelled = true;
        },
    };
}
