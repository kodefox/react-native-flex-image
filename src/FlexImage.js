// @flow

import React, {Component} from 'react';
import autobind from 'class-autobind';
import resolveAssetSource from 'react-native/Libraries/Image/resolveAssetSource';
import {
  View,
  Text,
  Image,
  ActivityIndicator,
  TouchableOpacity,
  Animated,
} from 'react-native';

type Cancellable = {
  cancel: () => void,
};

type Props = {
  source: number | { uri: string, width?: number, height?: number },
  style?: StyleType,
  loadingComponent?: ReactNode,
  onPress?: () => void,
  thumbnail?: number | { uri: string, width?: number, height?: number },
  loadingMethod?: 'spinner' | 'progressive',
};

type State = {
  isLoading: boolean,
  ratio: ?number,
  error: ?string,
  thumbnailOpacity: Animated.Value,
};

export default class FlexImage extends Component<Props, State> {
  _pendingGetSize: ?Cancellable;

  constructor(props: Props, ...args: Array<mixed>) {
    super(props, ...args);
    autobind(this);

    let {source, thumbnail, loadingMethod} = props;
    let ratio;
    let error;
    let isLoading = true;

    let src = thumbnail && loadingMethod === 'progressive' ? thumbnail : source;
    if (typeof src === 'number') {
      let imageSource = resolveAssetSource(src);
      if (imageSource) {
        let {width, height} = imageSource;
        if (width && height) {
          ratio = width / height;
        }
      } else {
        error = 'Error: Failed to retrieve width and height of the image';
      }
      isLoading = false;
    } else {
      this._pendingGetSize = getImageSize(
        src,
        this._onLoadSuccess,
        this._onLoadFail
      );
    }

    this.state = {
      ratio,
      isLoading,
      error,
      thumbnailOpacity: new Animated.Value(0),
    };
  }

  componentWillUnmount() {
    if (this._pendingGetSize) {
      this._pendingGetSize.cancel();
    }
  }

  render() {
    let {
      source,
      style,
      onPress,
      loadingComponent,
      thumbnail,
      loadingMethod,
      ...otherProps
    } = this.props;
    let {isLoading, ratio, error, thumbnailOpacity} = this.state;

    if (isLoading && loadingMethod !== 'progressive') {
      let loadingIndicator = loadingComponent || (
        <ActivityIndicator size="large" />
      );
      return (
        <View style={[{justifyContent: 'center', alignItems: 'center'}, style]}>
          {loadingIndicator}
        </View>
      );
    }

    if (error) {
      return (
        <View style={[{justifyContent: 'center', alignItems: 'center'}, style]}>
          <Text>{error}</Text>
        </View>
      );
    }

    let imageSource;
    if (typeof source === 'number') {
      imageSource = source;
    } else {
      // eslint-disable-next-line no-unused-vars
      let {uri, width, height, ...other} = source;
      imageSource = {uri, ...other};
    }

    return (
      <TouchableOpacity
        onPress={onPress}
        disabled={!onPress}
        style={[{aspectRatio: ratio}, style]}
      >
        {thumbnail &&
          loadingMethod === 'progressive' && (
            <Animated.Image
              {...otherProps}
              source={thumbnail}
              style={{
                width: '100%',
                height: '100%',
                opacity: thumbnailOpacity,
                zIndex: 1,
              }}
              onLoad={this._onThumbnailLoad}
              testID="progressiveThumbnail"
            />
          )}
        <Animated.Image
          {...otherProps}
          source={imageSource}
          style={{width: '100%', height: '100%', position: 'absolute'}}
          onLoad={this._onLoad}
        />
      </TouchableOpacity>
    );
  }

  _onThumbnailLoad = () => {
    Animated.timing(this.state.thumbnailOpacity, {
      toValue: 1,
      duration: 250,
    }).start();
  };

  _onLoad = () => {
    Animated.timing(this.state.thumbnailOpacity, {
      toValue: 0,
      duration: 250,
    }).start();
  };

  _onLoadSuccess(width: number, height: number) {
    let ratio = width / height;
    this.setState({
      isLoading: false,
      ratio,
    });
  }

  _onLoadFail(error: Error) {
    this.setState({
      isLoading: false,
      error: 'Error: ' + error.message,
    });
  }
}

// A cancellable version of Image.getSize
export function getImageSize(
  source: { uri: string },
  onSuccess: (width: number, height: number) => void,
  onFail: (error: Error) => void
) {
  let isCancelled = false;
  Image.getSize(
    source.uri,
    (width: number, height: number) => {
      if (!isCancelled) {
        onSuccess(width, height);
      }
    },
    (error: Error) => {
      if (!isCancelled) {
        onFail(error);
      }
    }
  );
  return {
    cancel: () => {
      isCancelled = true;
    },
  };
}
