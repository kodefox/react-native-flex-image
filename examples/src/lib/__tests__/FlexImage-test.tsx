import React from 'react';
import FlexImage, {getImageSize} from '../FlexImage';
import {render} from 'react-native-testing-library';

type Callback = (width: number, height: number) => void;

jest.mock('Image', () => ({
  getSize: jest.fn((_uri: string, callback: Callback) => {
    setTimeout(() => {
      callback(200, 300);
    }, 1);
  }),
}));

jest.mock('Animated', () => ({
  Image: 'Animated.Image',
  View: 'Animated.View',
  createTimer: jest.fn(),
  Value: class AnimatedValue {
    constructor(value: string) {
      return {
        interpolate: () => value,
        __getValue: () => value,
      };
    }
  },
  timing: jest.fn(() => ({
    start: (callback: Callback) => callback && callback(200, 300),
  })),
}));

it('should render thumbnail when loadingMethod is progressive', () => {
  let component = render(
    <FlexImage
      source={{
        uri: 'https://i.imgur.com/v0Y4Jfu.jpg',
      }}
      thumbnail={{uri: 'https://i.imgur.com/ngVArkT.jpg'}}
      loadingMethod="progressive"
    />,
  );
  expect(component.toJSON()).toMatchSnapshot();
});

it('should run onSuccess function when width and height are resolved', async () => {
  let onSuccess;
  let onFailed;
  await new Promise((resolve, reject) => {
    let source = {uri: 'imageURI'};
    onSuccess = jest.fn(resolve);
    onFailed = jest.fn(reject);
    getImageSize(source, onSuccess, onFailed);
  });
  expect(onSuccess).toHaveBeenCalledWith(200, 300);
  expect(onFailed).not.toHaveBeenCalled();
});

it('should not run onSuccess function when isCancelled is true', () => {
  let source = {uri: 'imageURI'};
  let onSuccess = jest.fn();
  let onFailed = jest.fn();
  let {cancel} = getImageSize(source, onSuccess, onFailed);
  cancel();
  expect(onSuccess).not.toHaveBeenCalled();
});
