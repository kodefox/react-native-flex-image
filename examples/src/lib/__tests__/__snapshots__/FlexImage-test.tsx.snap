// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render thumbnail when loadingMethod is progressive 1`] = `
<Animated.View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Array [
        Object {
          "aspectRatio": -1,
        },
        undefined,
      ],
      Object {
        "opacity": Object {
          "__getValue": [Function],
          "interpolate": [Function],
        },
      },
    ]
  }
>
  <Animated.Image
    onLoad={[Function]}
    source={
      Object {
        "uri": "https://i.imgur.com/ngVArkT.jpg",
      }
    }
    style={
      Object {
        "height": "100%",
        "opacity": Object {
          "__getValue": [Function],
          "interpolate": [Function],
        },
        "width": "100%",
        "zIndex": 1,
      }
    }
    testID="progressiveThumbnail"
  />
  <Animated.Image
    onLoad={[Function]}
    source={
      Object {
        "uri": "https://i.imgur.com/v0Y4Jfu.jpg",
      }
    }
    style={
      Object {
        "height": "100%",
        "position": "absolute",
        "width": "100%",
      }
    }
  />
</Animated.View>
`;
